metapool/sample_sheet.py

```diff
@@ -1488,42 +1524,99 @@ def __init__(self, path=None):
+    # column and raise an Error if not. By convention it should be, once
+    # legacy comments and whitespace rows are removed.
+    if df[0][0] != '[Header]':
+        raise ValueError("Top section is not [Header]")
 
-    sheet = MetagenomicSampleSheetv100(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # identify the beginning of the following section and remove everything
+    # from the start of the first section on down. Remove the now redundant
+    # [Header] from the top row as well.
+    next_section_start = df.index[df[0].str.startswith("[")].tolist()[1]
+    df = df.iloc[1:next_section_start]
 
-    sheet = MetagenomicSampleSheetv90(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # lastly, trim off the additional all-empty columns that are now
+    # present after the removal of the other sections.
+    df.dropna(how='all', axis=1, inplace=True)
 
-    sheet = MetatranscriptomicSampleSheetv10(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # set the index to the attributes column of the sample-sheet, replacing
+    # the numeric index which now isn't needed. The dataframe will now just
+    # contain the index column and a single column named 1.
+    df.set_index(0, inplace=True)
+
+    # return the value of key '1'. This will return an immediately
+    # recognizable dictionary of key/value pairs.
+    results = df.to_dict()[1]
+
+    # conversion to dict causes SheetVersion to be wrapped in single ticks.
+    # e.g.: "'100'". These should be removed if present.
+    if 'SheetVersion' in results:
+        results['SheetVersion'] = results['SheetVersion'].replace("'", "")
+
+    return results
+
+
+def load_sample_sheet(sample_sheet_path):
+    types = [AmpliconSampleSheet, MetagenomicSampleSheetv101,
+             MetagenomicSampleSheetv100, MetagenomicSampleSheetv90,
+             AbsQuantSampleSheetv10, MetatranscriptomicSampleSheetv0,
+             MetatranscriptomicSampleSheetv10]
+
+    header = _parse_header(sample_sheet_path)
+
+    required_attributes = ['Assay', 'SheetType', 'SheetVersion']
+    missing_attributes = []
+    for attribute in required_attributes:
+        if attribute not in header:
+            missing_attributes.append(f"'{attribute}'")
+
+    if len(missing_attributes) != 0:
+        raise ValueError("The following fields must be defined in [Header]: "
+                         " %s" % ", ".join(missing_attributes))
+
+    sheet = None
+
+    for type in types:
+        m = True
+        for attribute in ['SheetType', 'SheetVersion', 'Assay']:
+            if type._HEADER[attribute] != header[attribute]:
+                m = False
+                break
+
+        if m is True:
+            # header matches all the attributes for the type.
+            sheet = type(sample_sheet_path)
+            break
```

Comment:
It seems like there at most needs to be a single break from this nested for loop but perhaps I'm misunderstanding the intent of the code

Reply:
Punt @charles-cowart

Reply:
This break is to break out of the enclosing 'for type in types:' loop. The first break just breaks you out of the key/value pair check early once a single False is found, because a single False is already a deal-breaker. We want the second break so we can exit the 'for type in types:' loop once a single matching SampleSheet() type is found; this is because there should only be one and only one match and we don't want to iterate through the entire set of child SampleSheet() classes when we don't have to.

Reply:
Does the example code with a single break in the comment a few lines up not work?

Reply:
It's functionally identical, isn't it? The matches() function you proposed is equivalent to the inner loop.

Reply:
I don't know but it's has a lower cyclomatic complexity, which tends to relate both to the readability of block of code and the ease of testing it.